revscoring/dependencies/functions.py

```diff
@@ -0,0 +1,259 @@
+            pairs.  When these items are reached while scanning the tree,
+            "CACHED" will be printed.
+
+    :Returns:
+        None
+    """
+    return "\n".join(draw_lines(dependent, context, cache, depth)) \
+           + "\n"
+
+def draw_lines(dependent, context, cache, depth):
+    cache = cache or {}
+    context = normalize_context(context)
+
+
+    if dependent in cache:
+        yield "\t" * depth + " - " + str(dependent) + " CACHED"
+    else:
+        if dependent in context:
+            dependent = context[dependent]
+
+        yield "\t" * depth + " - " + str(dependent)
+
+        # Check if we're a dependent with explicit dependencies
+        if hasattr(dependent, "dependencies"):
+            for dependency in dependent.dependencies:
+                yield from draw_lines(dependency, context, cache, depth+1)
+
+def dig(dependents, context=None, cache=None):
+    """
+    Expands root dependencies.  These are dependents at the bottom of the tree
+    -- dependents with no dependencies of their own.
+
+    :Parameters:
+        dependents : `Dependent` | `iterable`
+            A dependent or collection of dependents to scan
+        context : `dict` | `iterable`
+            A mapping of injected dependency processers to use as context.
+            Can be specified as a set of new `Dependent` or a map of `Dependent`
+            pairs.
+        cache : `dict` | `set`
+            A cache of previously solved dependencies to not scan beneath
+
+    :Returns:
+        A generator over root dependencies
+    """
+    cache = set(cache or [])
+    context = normalize_context(context)
+
+    if hasattr(dependents, '__iter__'):
+        # Multiple values
+        return _dig_many(dependents, context, cache)
+    else:
+        # Singular value
+        dependent = dependents
+        return _dig(dependent, context, cache)
+
+
+def normalize_context(context):
+    """
+    Normalizes a context argument.  This allows for context to be specified
+    either as a collection of contextual `Dependant` or a `dict` of `Dependent`
+    pairs.
+    """
+    if context is None:
+        return {}
+    elif isinstance(context, dict):
+        return context
+    elif hasattr(context, "__iter__"):
+        return {d:d for d in context}
+    else:
+        raise TypeError("'context' is not a dict or iterable: {0}" \
+                        .format(str(context)))
+
+def _solve(dependent, context, cache, history=None):
+    history = history or set()
+
+    # Check if we've already got a value for this dependency
+    if dependent in cache:
+        return cache[dependent], cache, history
+
+    # Check if a corresponding dependent was injected into the context
+    else:
+
+        # If a dependent is in context here, replace it.
+        if dependent in context:
+            dependent = context[dependent]
+
+        # Check if the dependency is callable.  If not, we're SOL
```

Comment:
Yeah... that's a euphemism that probably doesn't belong in a comment.  http://www.urbandictionary.com/define.php?term=SOL